### PR TITLE
Implementation of an Artnet Protocol for controlling light fixtures.

### PR DESCRIPTION
**Pull request by RamonPeek**
Our protocol can be used to control light fixtures via Artnet over DMX. The user can create an ArtnetAgent in the manager, in which the network-details are defined and the user can import light assets via json-import.

The import-functionality has been overwritten to also support replacement, addition and removal of existing assets.

### CODE REVIEW

File: agent/src/main/java/org/openremote/agent/protocol/AbstractProtocol.java [hunk 1]
```diff
@@ -510,7 +511,7 @@ public String toString() {
     /**
      * Link an attribute to its linked protocol configuration.
      */
-    abstract protected void doLinkAttribute(AssetAttribute attribute, AssetAttribute protocolConfiguration);
+    abstract protected void doLinkAttribute(AssetAttribute attribute, AssetAttribute protocolConfiguration) throws IOException;
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/AbstractProtocol.java:514]: Please change to `Exception` not `IOException`, attributes shouldn't be establishing the connection that is the job of the `doLinkProtocolConfiguration` so shouldn't see `IOException` here in general so `throws` shouldn't be limited  to that type

File: agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetLightState.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package org.openremote.agent.protocol.dmx;
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetLightState.java:1]: Don't understand why some files are in the `dmx` package whilst others are in `dmx.artnet`

File: agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetLight.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
+package org.openremote.agent.protocol.dmx;
+
+public abstract class AbstractArtnetLight {
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetLight.java:3]: This is not an abstract class...there's nothing abstract in it....combine with `ArtnetLight`

File: model/build.gradle [hunk 4]
```diff
@@ -5,7 +5,7 @@ compileTestJava.options.compilerArgs.add '-parameters'
 
 dependencies {
     compile("com.google.gwt.inject:gin:$ginVersion")
-
+    compile "com.fasterxml.jackson.core:jackson-annotations:$jacksonVersion"
```
Comment (richturner) [inline model/build.gradle:8]: Why is this needed....existing `jackson-databind` has a dependency on this and model module compiles fine....please undo

File: manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java [hunk 5]
```diff
@@ -20,13 +20,15 @@
 package org.openremote.manager.setup.builtin;
 
 import org.apache.commons.io.IOUtils;
+import org.openremote.agent.protocol.dmx.artnet.ArtnetClientProtocol;
```
Comment (richturner) [inline manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java:23]: We don't want to provision `Artnet` protocol in our demo setup as it doesn't make sense, don't mind it here for dev purposes but will need to be removed from here....we don't provision protocols in demo as connections won't work

File: manager/src/main/java/org/openremote/manager/setup/builtin/BuiltinSetupTasks.java [hunk 6]
```diff
@@ -41,6 +41,7 @@
             addTask(new ManagerCleanSetup(container));
             addTask(new KeycloakCleanSetup(container));
             addTask(new KeycloakInitSetup(container));
+            addTask(new ArtNetDemoSetup());
```
Comment (richturner) [inline manager/src/main/java/org/openremote/manager/setup/builtin/BuiltinSetupTasks.java:44]: Don't want a fixed setup in our system for `Artnet` for reasons mentioned in `ManagerDemoSetup`

File: manager/src/main/java/org/openremote/manager/setup/builtin/ArtNetDemoSetup.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
+package org.openremote.manager.setup.builtin;
```
Comment (richturner) [inline manager/src/main/java/org/openremote/manager/setup/builtin/ArtNetDemoSetup.java:1]: No other protocol uses a setup task to provision assets and provisioning is project specific and is always handled in the project code

File: gradlew [hunk 8]
```diff
@@ -1,172 +0,0 @@
-#!/usr/bin/env sh
```
Comment (richturner) [inline gradlew:1]: Please undo this should not be deleted

File: agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetLightState.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+    public AbstractArtnetLightState(int lightId) {
+        this.lightId = lightId;
+    }
+    public int getLightId() {
+        return this.lightId;
+    }
+    public abstract Byte[] getValues();
+
+    public abstract void fromAttribute(AttributeEvent event, Attribute attribute);
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetLightState.java:17]: I only see 1 implementation of this abstract class....why not make it a proper class?

File: agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetClientProtocol.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+import static org.openremote.model.Constants.PROTOCOL_NAMESPACE;
+import static org.openremote.model.attribute.MetaItemDescriptor.Access.ACCESS_PRIVATE;
+import static org.openremote.model.attribute.MetaItemDescriptorImpl.metaItemInteger;
+import static org.openremote.model.syslog.SyslogCategory.PROTOCOL;
+
+public abstract class AbstractArtnetClientProtocol<T> extends AbstractIoClientProtocol<T, UdpIoClient<T>> {
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/AbstractArtnetClientProtocol.java:20]: Not sure why this is abstract....should just be:

```public class ArtnetClientProtocol extends AbstractIoClientProtocol<ArtnetPacket, UdpIoClient<ArtnetPacket>```

File: agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+import static org.openremote.model.asset.AssetType.THING;
+import static org.openremote.model.attribute.AttributeValueType.*;
+import static org.openremote.model.attribute.MetaItemDescriptor.Access.ACCESS_PRIVATE;
+import static org.openremote.model.attribute.MetaItemDescriptorImpl.metaItemInteger;
+import static org.openremote.model.attribute.MetaItemDescriptorImpl.metaItemObject;
+import static org.openremote.model.attribute.MetaItemType.AGENT_LINK;
+import static org.openremote.model.attribute.MetaItemType.READ_ONLY;
+import static org.openremote.model.syslog.SyslogCategory.PROTOCOL;
+
+public class ArtnetClientProtocol extends AbstractArtnetClientProtocol<ArtnetPacket> implements ProtocolLinkedAttributeImport {
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:47]: As mentioned in `AbstractArtnetClientProtocol` this should be combined

File: agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+                    consumers = new ArrayList<>();
+                }
+                consumers.add(new Pair<AttributeRef, Consumer<ArtnetPacket>>(
+                        attribute.getReferenceOrThrow(),
+                        messageConsumer
+                ));
+                return consumers;
+            });
+        }
+        if(getLinkedAttribute(attribute.getReference().orElse(null)) != null) {
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:160]: Nested `if` statements are best avoided if possible, prefer `if not then return`

File: agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+                    Attribute universeAttribute = parentAsset.getAttribute("Universe").orElse(null);
+                    Attribute lightAttribute = parentAsset.getAttribute("Id").orElse(null);
+                    if(universeAttribute != null && lightAttribute != null) {
+                        int universeId = universeAttribute.getValueAsInteger().orElse(-1);
+                        int lightId = lightAttribute.getValueAsInteger().orElse(-1);
+                        if(universeId != -1 && lightId != -1) {
+                            ArtnetLight updatedLight = (ArtnetLight) artnetLightMemory.stream().filter(light -> light.getLightId() == lightId).findFirst().orElse(null);
+                            if(updatedLight != null) {
+                                ArtnetLightState oldLightState = (ArtnetLightState) updatedLight.getLightState();
+                                ObjectMapper mapper = new ObjectMapper();
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:259]: Use `Container.JSON`

Comment (richturner) [reply agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:259]: There are also helper methods in `Values` `convertFromValue` and `convertToValue`

File: agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java [hunk 14]
```diff
@@ -0,0 +1,502 @@
...
+            }));
+            return encodersDecoders.toArray(new ChannelHandler[0]);
+        };
+        return encoderDecoderProvider;
+    }
+
+    @Override
+    protected void doLinkAttribute(AssetAttribute attribute, AssetAttribute protocolConfiguration) {
+        AttributeRef protocolRef = protocolConfiguration.getReferenceOrThrow();
+        Consumer<ArtnetPacket> messageConsumer = artnetPacket -> {};
```
Comment (richturner) [inline agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:147]: What's the purpose of this empty consumer? How are linked attributes updated when a change occurs to a light for example?

File: agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java (hunk 13 above)
Comment (RamonPeek) [reply agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:259]: > There are also helper methods in `Values` `convertFromValue` and `convertToValue`

We have chosen to check each individual field in the json-object to make sure whether it is present or not. This is important as we do not want to update the lightstate if the provided json is invalid (is missing a property).

Comment (richturner) [reply agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java:259]: Use convertFromValue to get the lightstate object then check that object for validity

**Comment by RamonPeek** [comment]
We have changed our protocol based on your feedback, please let us know when we need to change anything else.

**Comment by RamonPeek** [comment]
Pulled latest changes from openremote:master and our commits from or-dmx fork. Everything should now be included

File: model/build.gradle [hunk 15]
```diff
@@ -5,7 +5,6 @@ compileTestJava.options.compilerArgs.add '-parameters'
 
 dependencies {
     compile("com.google.gwt.inject:gin:$ginVersion")
```
Comment (richturner) [inline model/build.gradle:7]: Please revert as it was before

File: manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java [hunk 16]
```diff
@@ -74,6 +74,7 @@
     public static GeoJSONPoint AREA_3_LOCATION = new GeoJSONPoint(5.487478, 51.446979);
     public static final String agentProtocolConfigName = "simulator123";
     public static final String thingLightToggleAttributeName = "light1Toggle";
+
```
Comment (richturner) [inline manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java:77]: Please revert as it was before

File: gradlew [hunk 17]
```diff
@@ -76,14 +76,12 @@ if [ -n "$JAVA_HOME" ] ; then
     fi
```
Comment (richturner) [inline gradlew:76]: Please revert as it was before

Comment (RamonPeek) [reply gradlew:76]: It hasn't been removed. It only swapped place (See line 9). I guess this has something to do with pulling the latest version of master.

File: manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java (hunk 16 above)
Comment (RamonPeek) [reply manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java:77]: What exactly needs to be reverted here? I think this was commited in the latest version of the openremote:master, since this change happened when I pulled the latest version into our fork.

File: model/build.gradle (hunk 15 above)
Comment (RamonPeek) [reply model/build.gradle:7]: What exactly should be reverted? Only change I see is the removal of a whitespace which was done automatically whilst pulling the latest version of master.

File: gradlew (hunk 17 above)
Comment (richturner) [reply gradlew:76]: Your file should match what is in master I shouldn't see any changes on this file compared to master.

File: manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java (hunk 16 above)
Comment (richturner) [reply manager/src/main/java/org/openremote/manager/setup/builtin/ManagerDemoSetup.java:77]: Your file should match what is in master I shouldn't see any changes on this file compared to master.

File: model/build.gradle (hunk 15 above)
Comment (richturner) [reply model/build.gradle:7]: Your file should match what is in master I shouldn't see any changes on this file compared to master.

**Comment by RamonPeek** [comment]
Reverted changes in the 3 files so they are the same as master
